File: src/lib/modelPlatform.ts
import { invoke } from "@tauri-apps/api/core";

export type CatalogSourceKind = "builtin" | "runtime";

export interface CapabilityFlags {
  downloadable: boolean;
  loadable: boolean;
  local_only: boolean;
  supports_translation: boolean;
  supports_streaming: boolean;
  supports_voice_cloning: boolean;
  supports_instruction_prompt: boolean;
  supports_inline_tags: boolean;
}

export interface RuntimeRequirement {
  id: string;
  label: string;
  engine_family: string;
  auto_routed: boolean;
}

export type TtsExpressivenessMode = "native" | "mapped" | "unsupported";

export type TtsAdvancedControlKind = "slider" | "toggle" | "select" | "text";
export type TtsControlGroup =
  | "identity"
  | "tempo"
  | "style"
  | "sampler"
  | "guidance"
  | "steering";

export interface TtsAdvancedControlOption {
  value: string;
  label: string;
}

export interface TtsAdvancedControlDescriptor {
  id: string;
  group: TtsControlGroup;
  label: string;
  description?: string | null;
  kind: TtsAdvancedControlKind;
  min?: number | null;
  max?: number | null;
  step?: number | null;
  unit?: string | null;
  options: TtsAdvancedControlOption[];
  default_value?:
    | { kind: "number"; value: number }
    | { kind: "boolean"; value: boolean }
    | { kind: "text"; value: string }
    | null;
}

export interface TtsDeliverySupport {
  expressiveness_mode: TtsExpressivenessMode;
  advanced_controls: TtsAdvancedControlDescriptor[];
}

export interface ProviderDescriptor {
  id: string;
  domain: "stt" | "tts" | "llm";
  source_kind: CatalogSourceKind;
  label: string;
  description: string;
  source_label: string;
  source_url?: string | null;
  runtime: RuntimeRequirement;
  available: boolean;
  local_only: boolean;
  license_label?: string | null;
  capabilities: CapabilityFlags;
}

export interface CatalogModelDescriptor {
  id: string;
  provider_id: string;
  domain: "stt" | "tts" | "llm";
  source_kind: CatalogSourceKind;
  label: string;
  description: string;
  installed: boolean;
  selected: boolean;
  active: boolean;
  runnable: boolean;
  downloadable: boolean;
  source_label: string;
  source_url?: string | null;
  runtime: RuntimeRequirement;
  license_label?: string | null;
  locale?: string | null;
  supported_languages: string[];
  readiness_status?: string | null;
  readiness_issues: string[];
  capabilities: CapabilityFlags;
  delivery_support: TtsDeliverySupport;
}

export interface DomainCatalog {
  providers: ProviderDescriptor[];
  models: CatalogModelDescriptor[];
}

export interface ModelPlatformSelectionState {
  selected_stt_provider_id?: string | null;
  selected_stt_model_id?: string | null;
  selected_llm_provider_id?: string | null;
  selected_llm_model_id?: string | null;
  selected_tts_provider_id?: string | null;
  selected_tts_model_id?: string | null;
  selected_tts_voice_id?: string | null;
  selected_tts_profile_id?: string | null;
  active_tts_provider_id?: string | null;
  active_tts_model_id?: string | null;
}

export interface ModelPlatformOverview {
  stt: DomainCatalog;
  llm: DomainCatalog;
  tts: DomainCatalog;
  selection: ModelPlatformSelectionState;
}

function parseInvokeJson<T>(command: string, json: string): T {
  try {
    return JSON.parse(json) as T;
  } catch (error) {
    const message =
      error instanceof Error ? error.message : "Invalid JSON returned.";
    throw new Error(`${command} returned malformed data: ${message}`);
  }
}

export async function getModelPlatformOverview(): Promise<ModelPlatformOverview> {
  const json = await invoke<string>("get_model_platform_overview_json");
  return parseInvokeJson<ModelPlatformOverview>(
    "get_model_platform_overview_json",
    json,
  );
}

export async function setTtsPlatformSelection(
  providerId: string,
  modelId?: string | null,
): Promise<void> {
  return invoke("set_tts_platform_selection", {
    providerId,
    modelId: modelId ?? null,
  });
}

File: src-tauri/src/model_platform.rs
use crate::external_model_storage::ModelStorageLocation;
use crate::settings::TtsStyleControlValue;
use serde::{Deserialize, Serialize};
use specta::Type;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Type)]
#[serde(rename_all = "snake_case")]
pub enum ModelDomain {
    Stt,
    Tts,
    Llm,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Type)]
#[serde(rename_all = "snake_case")]
pub enum CatalogSourceKind {
    Builtin,
    Runtime,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
pub struct CapabilityFlags {
    pub downloadable: bool,
    pub loadable: bool,
    pub local_only: bool,
    pub supports_translation: bool,
    pub supports_streaming: bool,
    pub supports_voice_cloning: bool,
    pub supports_instruction_prompt: bool,
    pub supports_inline_tags: bool,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Type)]
#[serde(rename_all = "snake_case")]
pub enum TtsExpressivenessMode {
    Native,
    Mapped,
    Unsupported,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Type)]
#[serde(rename_all = "snake_case")]
pub enum TtsAdvancedControlKind {
    Slider,
    Toggle,
    Select,
    Text,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Type)]
#[serde(rename_all = "snake_case")]
pub enum TtsControlGroup {
    Identity,
    Tempo,
    Style,
    Sampler,
    Guidance,
    Steering,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
pub struct TtsAdvancedControlOption {
    pub value: String,
    pub label: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
pub struct TtsAdvancedControlDescriptor {
    pub id: String,
    pub group: TtsControlGroup,
    pub label: String,
    pub description: Option<String>,
    pub kind: TtsAdvancedControlKind,
    pub min: Option<f32>,
    pub max: Option<f32>,
    pub step: Option<f32>,
    pub unit: Option<String>,
    pub options: Vec<TtsAdvancedControlOption>,
    pub default_value: Option<TtsStyleControlValue>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
pub struct TtsDeliverySupport {
    pub expressiveness_mode: TtsExpressivenessMode,
    pub advanced_controls: Vec<TtsAdvancedControlDescriptor>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
pub struct RuntimeRequirement {
    pub id: String,
    pub label: String,
    pub engine_family: String,
    pub auto_routed: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
pub struct ProviderDescriptor {
    pub id: String,
    pub domain: ModelDomain,
    pub source_kind: CatalogSourceKind,
    pub label: String,
    pub description: String,
    pub source_label: String,
    pub source_url: Option<String>,
    pub runtime: RuntimeRequirement,
    pub available: bool,
    pub local_only: bool,
    pub license_label: Option<String>,
    pub capabilities: CapabilityFlags,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
pub struct CatalogModelDescriptor {
    pub id: String,
    pub provider_id: String,
    pub domain: ModelDomain,
    pub source_kind: CatalogSourceKind,
    pub label: String,
    pub description: String,
    pub installed: bool,
    pub selected: bool,
    pub active: bool,
    pub runnable: bool,
    pub downloadable: bool,
    pub source_label: String,
    pub source_url: Option<String>,
    pub runtime: RuntimeRequirement,
    pub license_label: Option<String>,
    pub locale: Option<String>,
    pub supported_languages: Vec<String>,
    pub readiness_status: Option<String>,
    pub readiness_issues: Vec<String>,
    pub capabilities: CapabilityFlags,
    pub delivery_support: TtsDeliverySupport,
    #[serde(default)]
    pub storage_location: Option<ModelStorageLocation>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
pub struct DomainCatalog {
    pub providers: Vec<ProviderDescriptor>,
    pub models: Vec<CatalogModelDescriptor>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
pub struct ModelPlatformSelectionState {
    pub selected_stt_provider_id: Option<String>,
    pub selected_stt_model_id: Option<String>,
    pub selected_llm_provider_id: Option<String>,
    pub selected_llm_model_id: Option<String>,
    pub selected_tts_provider_id: Option<String>,
    pub selected_tts_model_id: Option<String>,
    pub selected_tts_voice_id: Option<String>,
    pub selected_tts_profile_id: Option<String>,
    pub active_tts_provider_id: Option<String>,
    pub active_tts_model_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
pub struct ModelPlatformOverview {
    pub stt: DomainCatalog,
    pub llm: DomainCatalog,
    pub tts: DomainCatalog,
    pub selection: ModelPlatformSelectionState,
}
